runtime/drivers/sqlite/catalog.go

```diff
@@ -140,3 +141,203 @@ func (c *catalogStore) DeleteResources(ctx context.Context) error {
+
+	var res []drivers.ModelSplit
+	for rows.Next() {
+		var elapsedMs int64
+		r := drivers.ModelSplit{}
+		err := rows.Scan(&r.Key, &r.DataJSON, &r.Index, &r.Watermark, &r.ExecutedOn, &r.Error, &elapsedMs)
+		if err != nil {
+			return nil, err
+		}
+		r.Elapsed = time.Duration(elapsedMs) * time.Millisecond
+		res = append(res, r)
+	}
+
+	if rows.Err() != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (c *catalogStore) FindModelSplitsByPending(ctx context.Context, modelID string, limit int) ([]drivers.ModelSplit, error) {
+	rows, err := c.db.QueryxContext(
+		ctx,
+		"SELECT key, data_json, idx, watermark, executed_on, error, elapsed_ms FROM model_splits WHERE instance_id=? AND model_id=? AND executed_on IS NULL ORDER BY idx LIMIT ?",
+		c.instanceID,
+		modelID,
+		limit,
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var res []drivers.ModelSplit
+	for rows.Next() {
+		var elapsedMs int64
+		r := drivers.ModelSplit{}
+		err := rows.Scan(&r.Key, &r.DataJSON, &r.Index, &r.Watermark, &r.ExecutedOn, &r.Error, &elapsedMs)
+		if err != nil {
+			return nil, err
+		}
+		r.Elapsed = time.Duration(elapsedMs) * time.Millisecond
+		res = append(res, r)
+	}
+
+	if rows.Err() != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (c *catalogStore) CheckModelSplitsHaveErrors(ctx context.Context, modelID string) (bool, error) {
+	rows, err := c.db.QueryContext(
+		ctx,
+		"SELECT 1 FROM model_splits WHERE instance_id=? AND model_id=? AND error != '' LIMIT 1",
+		c.instanceID,
+		modelID,
+	)
+	if err != nil {
+		return false, err
+	}
+	defer rows.Close()
+
+	var hasErrors bool
+	if rows.Next() {
+		hasErrors = true
+	}
+
+	if rows.Err() != nil {
+		return false, err
+	}
+
+	return hasErrors, nil
+}
+
+func (c *catalogStore) InsertModelSplit(ctx context.Context, modelID string, split drivers.ModelSplit) error {
+	_, err := c.db.ExecContext(
+		ctx,
+		"INSERT INTO model_splits(instance_id, model_id, key, data_json, idx, watermark, executed_on, error, elapsed_ms) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
+		c.instanceID,
+		modelID,
+		split.Key,
+		split.DataJSON,
+		split.Index,
+		split.Watermark,
+		split.ExecutedOn,
+		split.Error,
+		split.Elapsed.Milliseconds(),
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *catalogStore) UpdateModelSplit(ctx context.Context, modelID string, split drivers.ModelSplit) error {
+	_, err := c.db.ExecContext(
+		ctx,
+		"UPDATE model_splits SET data_json=?, idx=?, watermark=?, executed_on=?, error=?, elapsed_ms=? WHERE instance_id=? AND model_id=? AND key=?",
+		split.DataJSON,
+		split.Index,
+		split.Watermark,
+		split.ExecutedOn,
+		split.Error,
+		split.Elapsed.Milliseconds(),
+		c.instanceID,
+		modelID,
+		split.Key,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdateModelSplitsPendingIfError(ctx context.Context, modelID string) error
+func (c *catalogStore) UpdateModelSplitsPendingIfError(ctx context.Context, modelID string) error {
```

Comment:
Seems unused and misplaced doc

Reply:
Thanks, fixed the docstring.

I added this function to support triggering refresh of all errored splits, but decided to implement support for the triggers in a separate PR. So I'll keep this func here and will use it in a follow up :)

Reply:
Thanks, fixed the docstring.

I added this function to support triggering refresh of all errored splits, but decided to implement support for the triggers in a separate PR. So I'll keep this func here and will use it in a follow up :)